### PR TITLE
Propose a typedefs header for paddle framework

### PR DESCRIPTION
**Pull request by reyoung**
The original discuess in

* https://github.com/PaddlePaddle/Paddle/pull/2548#discussion_r123270414
* https://github.com/PaddlePaddle/Paddle/pull/2579#discussion_r123676641

This commit is just a proposal, let's do such kind of summarization in
this PR.

### CODE REVIEW

**Comment by reyoung** [comment]
I just run

```bash
grep std (find . -name '*.h' -o -name '*.cpp') | sed -r 's#.*std\:\:([a-z_0-9]+)[<(\ ].*#\1#g' | grep -v 'std::' | sort | uniq -c | sort -r | head -n 10
```

in current PaddlePaddle repo. It returns

```text
1116 vector
 601 string
 289 make_shared
 278 unique_ptr
 140 shared_ptr
  82 function
  76 setw
  75 dynamic_pointer_cast
  70 pair
  63 min
```

Maybe we should typedef:

* map/set  because the `unordered_map/unordered_set` is faster than default map/set. And `unordered_` is too long.
* unique_ptr/shared_ptr/make_unique(which in c++ 14 standard, but can be implemented manually)/make_shared/dynamic_pointer_cast because it is heavily used and there are many smart pointer implementations.
* vector/string/function/pair because they are heavily used.

And there are some classes in `boost` are very useful.

* boost::variant and boost::any. They are in C++ 17 standard.
* boost::non_copyable.  Mark a class is non-copyable. Instead of using a macro, using this base class would not pollute the global namespace.
* `boost::small_vector`. small_vector is a vector with a size limit N. If vector.size() < N, then all data will be allocated in stack space. And if vector.size() > N, it will allocate head data. It is very useful to represent `Op::inputs_` and `Op::outputs_`. Because each op could have a few inputs and outputs, e.g. 3 inputs, and 1 outputs. Use `small_vector` to hold `inputs` and `outputs` variable will make configuration of `Op` faster.

File: paddle/framework/typedefs.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+using WeakPtr = std::weak_ptr<T>;
+
+//! MakeShared will create std::shared_ptr
+template <typename T, typename... ARGS>
+inline SharedPtr<T> MakeShared(ARGS... args) {
+  return std::make_shared<T>(std::forward<ARGS>(args)...);
+}
+
+//! MakeUnique will create std::unique_ptr
+template <typename T, typename... ARGS>
```
Comment (gangliao) [inline paddle/framework/typedefs.h]: So you did not want to distinguish special case, for instance, is_array<T>::value

Comment (gangliao) [inline paddle/framework/typedefs.h]: ```auto a_uptr = make_unique<int[]>(10);```

File: paddle/framework/typedefs.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#include <unordered_set>
+#include <vector>
+
+namespace paddle {
+namespace framework {
+
+//! Using unordered_map as Paddle's default map. It will faster than
+//! std::map
+template <typename Key, typename Val>
+using Map = std::unordered_map<Key, Val>;
```
Comment (Canpio) [inline paddle/framework/typedefs.h]: How about `UMap` or `uMap`? `Map` really looks like the alias of `std::map`.

**Pull request closed without merging**